SSA/Experimental/Bits/Fast/BitStream.lean

```diff
@@ -373,12 +373,41 @@ variable {w : Nat} {x y : BitVec w} {a b a' b' : BitStream}
 
 local infix:20 " ≈ʷ " => EqualUpTo w
 
--- TODO: These sorries are difficult, and will be proven in a later Pull Request.
+-- TODO: This sorry is difficult, and will be proven in a later Pull Request.
 theorem ofBitVec_sub : ofBitVec (x - y) ≈ʷ (ofBitVec x) - (ofBitVec y)  := by
   sorry
 
+theorem Nat.two_mod_mod (x : Nat) (w : Nat) (_ : 0 < w) : x % 2 ^ w % 2 = x % 2 := by
+  have y : 2 ^ 1 ∣ 2 ^ w := Nat.pow_dvd_pow 2 (by omega)
+  rw [pow_one] at y
+  exact Nat.mod_mod_of_dvd x y
+
+theorem Nat.parity_and (n m : Nat) :
+    2 ≤ n % 2 + m % 2 ↔ n % 2 = 1 ∧ m % 2 = 1 := by
+  omega
+
+theorem Nat.add_odd_iff_neq (n m : Nat) :
+    (n + m) % 2 = 1 ↔ (n % 2 = 1) ≠ (m % 2 = 1) := by
+  cases Nat.mod_two_eq_zero_or_one n
+  <;> cases Nat.mod_two_eq_zero_or_one m
+  <;> (
+    rename_i mparity nparity
+    simp [mparity, nparity, Nat.add_mod]
```

Comment:
I would prefer `simp_all [Nat.add_mod]` than a `rename_i` then a `simp`

Reply:
Note that `simp[*, Nat.add_mod]` also exists (`simp_all` is more like `simp[*] at *`, i.e., the former will use all local hypotheses to simplify the goal only, while the latter will simplify everywhere).

I agree that `rename_i` is a bit ugly, and brittle!